core/src/main/java/com/linecorp/armeria/client/limit/AsyncConcurrencyLimit.java

```diff
@@ -0,0 +1,254 @@
+import static java.lang.Integer.MAX_VALUE;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+
+import io.netty.channel.EventLoop;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.FutureListener;
+import io.netty.util.concurrent.Promise;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+public class AsyncConcurrencyLimit implements ConcurrencyLimit<ClientRequestContext> {
+    /**
+     * Returns a new {@link ConcurrencyLimitBuilder} with the specified {@code maxConcurrency}.
+     *
+     * @param maxConcurrency the maximum number of concurrent active requests,
+     *                       specify {@code 0} to disable the limit.
+     */
+    public static ConcurrencyLimitBuilder builder(int maxConcurrency) {
+        checkArgument(maxConcurrency >= 0,
+                      "maxConcurrency: %s (expected: >= 0)", maxConcurrency);
+        return new ConcurrencyLimitBuilder(maxConcurrency == MAX_VALUE ? 0 : maxConcurrency);
+    }
+
+    /**
+     * Returns a new {@link ConcurrencyLimitBuilder} with the specified {@code maxConcurrency}.
+     *
+     * @param maxConcurrency the maximum number of concurrent active requests,
+     *                       specify {@code 0} to disable the limit.
+     */
+    public static ConcurrencyLimitBuilder of(int maxConcurrency) {
+        return builder(maxConcurrency);
+    }
+
+    private final int maxParallelism;
+    private final int maxPendingAcquires;
+    private final long acquireTimeoutNanos;
+
+    private final Queue<AcquireTask> pendingAcquireQueue;
+    private final AtomicInteger acquiredPermitCount = new AtomicInteger();
+    private final Runnable timeoutTask;
+    private final Permit semaphorePermit = this::decrementAndRunTaskQueue;
+    private int pendingAcquireCount;
+
+    AsyncConcurrencyLimit(
+            long acquireTimeoutMillis,
+            int maxParallelism,
+            int maxPendingAcquires) {
+        this.maxParallelism = maxParallelism;
+        this.maxPendingAcquires = maxPendingAcquires;
+        this.pendingAcquireQueue = new ArrayDeque<>(maxPendingAcquires);
+        this.acquireTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(acquireTimeoutMillis);
+        this.timeoutTask = new TimeoutTask();
+    }
+
+    /**
+     * Returns the total number of acquired permits.
+     */
+    public int acquiredPermitCount() {
+        return acquiredPermitCount.get();
+    }
+
+    /**
+     * Returns the total number of available permits.
+     */
+    public int availablePermitCount() {
+        return maxParallelism - acquiredPermitCount();
+    }
```

Comment:
How about moving these methods to `ConcurrencyLimit` and updating the contract a little bit where a return value `-1`  means 'unknown', so that users have access to these properties without downcasting while allowing some degree of freedom to the implementors?